client/me/purchases/confirm-cancel-purchase/index.jsx

```diff
@@ -24,43 +24,36 @@ const ConfirmCancelPurchase = React.createClass( {
 		selectedSite: React.PropTypes.object
 	},
 
-	componentWillMount() {
-		recordPageView( 'confirm_cancel_purchase', this.props );
+	getInitialState() {
+		return {
+			html: null
+		};
 	},
 
-	componentWillReceiveProps( nextProps ) {
-		recordPageView( 'confirm_cancel_purchase', this.props, nextProps );
+	componentWillMount() {
+		this.loadEndpointForm();
+		recordPageView( 'confirm_cancel_purchase', this.props );
 	},
 
-	componentDidMount() {
-		if ( ! this.props.selectedPurchase.hasLoadedFromServer || ! this.props.selectedSite ) {
-			return;
+	componentDidUpdate( prevProps, prevState ) {
+		if ( ! prevState.html && this.state.html ) {
+			const container = React.findDOMNode( this.refs.root );
+
+			container.innerHTML = this.state.html;
+			initializeForm( {
+				form: container.querySelector( 'form' ),
+				onSubmit: this.handleSubmit,
+				selectedPurchase: getPurchase( this.props ),
+				selectedSite: this.props.selectedSite
+			} );
 		}
-
-		this.resetEndpointForm();
 	},
 
-	componentDidUpdate( prevProps ) {
-		if ( ! this.props.selectedPurchase.hasLoadedFromServer ) {
-			return;
-		}
-
-		if ( isEqual( prevProps.selectedPurchase.data, this.props.selectedPurchase.data ) &&
-				( prevProps.selectedSite.ID === this.props.selectedSite.ID ) ) {
-			return;
-		}
-
-		this.resetEndpointForm();
-	},
-
-	resetEndpointForm() {
-		const { selectedSite, selectedPurchase } = this.props;
-
-		loadEndpointForm( {
-			container: React.findDOMNode( this.refs.root ),
-			selectedSite,
-			selectedPurchase: selectedPurchase.data,
-			onSubmit: this.handleSubmit
+	loadEndpointForm() {
+		loadEndpointForm( getPurchase( this.props ), ( html ) => {
+			this.setState( {
+				html
+			} );
```

Comment:
If we don't want to use the HTML in state in render(), this might be more appropriate as a boolean (`isFormLoaded` or something, which is all render() is concerned with) with the HTML injected in a callback:

```js
this.setState( { isFormLoaded: true }, () => {
  container.innerHTML = html;
  initializeForm( { ... } );
} );
```

My rationale:

- This uses the minimal state to represent render().
- We only want to inject the form once, so this is a little more explicit than using `componentDidUpdate` with a probably-one-time condition like `! prevState.html && this.state.html`.

Of course, this is all moot if we were to use `this.state.html` in render(). Apologies if I missed a reason this method should be avoided.